flips/sync-protocol.md

```diff
@@ -0,0 +1,109 @@
+* A Range Request response should contain a single chain of blocks which begins at the start height of the requested range and is no longer than the size of the requested range.
+* A Batch Request response should contain a subset of the requested block IDs.
+
+This eliminates any ambiguity about whether a response corresponds to a Batch or Range Request, so we can avoid optimistically setting the statuses of heights as Received for responses to Batch Requests.
+
+At any time, there is a single range of heights that the Sync Engine actively requests, which is tracked by the Sync Core. We call this the Active Range. The Active Range is parameterized by two variables `RangeStart` and `RangeEnd`, which effectively replace the `Heights` map from the existing implementation, but it can be broken up and requested by the Sync Engine in multiple segments. `RangeStart` should be greater than the local finalized block height, and `RangeEnd` should be less than or equal to the target finalized block height (more details below). The logic for updating the Active Range can be abstracted with an interface:
+
+```golang
+type ActiveRange interface {
+    // Update processes a range of blocks received from a Range Request 
+    // response and updates the requestable height range.
+    Update(headers []flow.Header, originID flow.Identifier)
+
+    // LocalFinalizedHeight is called to notify a change in the local finalized height.
+    LocalFinalizedHeight(height uint64)
+
+    // TargetFinalizedHeight is called to notify a change in the target finalized height.
+    TargetFinalizedHeight(height uint64)
+
+    // Get returns the range of requestable block heights.
+    Get() flow.Range
+}
+```
+
+There are many ways to implement this interface, but one possible approach is as follows:
+* Select values for parameters `DefaultRangeSize` and `MinResponses`
+* Let `PendingStart` be the first height greater than `LocalFinalizedHeight` that has been received less than `MinResponses` times
+* Let `RangeStart` be equal to `LocalFinalizedHeight + 1`
+* Let `RangeEnd` be the smaller of `TargetFinalizedHeight` and `PendingStart + DefaultRangeSize`
+
+The reason we keep track of `PendingStart` is to ensure that `RangeEnd` eventually increases even if the local finalized height doesn't. This is needed to address the last item in [Potential Problems](#potential-problems). 
+
+The target finalized height represents the speculated finalized block height of the overall chain, and should reflect the Sync Height Responses that have been received while accounting for the possibility that some of these responses are malicious. Therefore, the Sync Height Response processing logic should incorporate some sort of expiration / filtering mechanism. The details of this logic can be abstracted with an interface:
+
+```golang
+type TargetFinalizedHeight interface {
```

Comment:
Does the sync core maintain a range of `TargetFinalizedHeight`, such that if say `ActiveRange.LocalFinalizedHeight` and `ActiveRange.TargetFinalizedHeight` has 100 heights apart, there are `100` `TargetFinalizedHeight` instances?
Does `TargetFinalizedHeight` maintain the `MinResponse`?

Reply:
there is only a single target finalized height. It is a single value that is a function of all the data points we've received from Sync Height Responses so far.

> Does TargetFinalizedHeight maintain the MinResponse

I don't really think `TargetFinalizedHeight` and `MinResponse` have that much to do with each other.